In [1]:
import nltk
import re

In [4]:
"""" Basic Regular Expression Meta-Characters, Including Wildcards, Ranges and Closures
.	        Wildcard, matches any character
^abc	    Matches some pattern abc at the start of a string
abc$	    Matches some pattern abc at the end of a string
[abc]	    Matches one of a set of characters
[^abc]      Matches anything but a set of characters
[A-Z0-9]	Matches one of a range of characters
ed|ing|s	Matches one of the specified strings (disjunction)
*	        Zero or more of previous item, e.g. a*, [a-z]* (also known as Kleene Closure)
+	        One or more of previous item, e.g. a+, [a-z]+
?	        Zero or one of the previous item (i.e. optional), e.g. a?, [a-z]?
{n}	        Exactly n repeats where n is a non-negative integer
{n,}	    At least n repeats
{,n}	    No more than n repeats
{m,n}	    At least m and no more than n repeats
a(b|c)+	    Parentheses that indicate the scope of the operators

SyntaxError: EOF while scanning triple-quoted string literal (<ipython-input-4-9b8a933661d3>, line 16)


# regular expression 
* list가 아니라 string에만 적용된다.
* ( ) 안에 들어 있는 것이 입력


In [ ]:
re.search('abc', 'abced')

* 입력이 2개인 함수이다.
* abcde라는 것에서 abc를 찾아라.

In [ ]:
wordlist = [w for w in nltk.corpus.words.words('en') if w.islower()]
print(wordlist[:10])


* wordlist는 string들의 list
* re.search는 list가 아니라 string에만 적용되기 때문에 for루프를 돌려서 하나씩 넣어주어야 한다.

In [ ]:
result = [w for w in wordlist if re.search('ed$', w)][:10]
print(result[:10])

* '~$'은 ~으로 끝나는 것을 의미한다.
* 이 경우 ed로 끝나는 단어를 앞에서부터 10개 보여주라는 의미

In [ ]:
result = [w for w in wordlist if re.search('^..j..t..$', w)][:10]
print(result[:10]

* .에는 어떤 글자라도 들어갈 수 있다.
* 이 경우 3번째 글자가 j, 5번째 글자가 t인 8글자의 단어를 앞에서부터 10개 보여주라는 의미

In [ ]:
result = [w for w in wordlist if re.search('^[ghi][mno][jlk][def]$', w)][:10]
print(result[:10])

* []은 안에 들어간 것 중에서 하나를 의미한다. []으로 묶으면 그 중 무엇이든 상관 없다는 뜻
* matches one of a set of characters. 
* 이 경우에는 g/h/i중에 하나, m/n/o중에 하나, j/l/k중에 하나, d/e/f중에 하나 인 4자리 글자를 보여주면 된다.
* 앞에서부터 10개 표시하라고 하였으나, 조건에 해당하는 글자가 4개뿐이라 4개만 나옴.

In [ ]:
result = [w for w in wordlist if re.search('^[ah]+$', w)][:10]
print(result[:10])

* +는 해당 글자가 한 번 이상 반복되는 것을 의미한다.
* 이 경우에는 a/h가 한 번 이상 반복되는 것을 보여주라는 의미.

In [ ]:
wsj = sorted(set(nltk.corpus.treebank.words()))
wordlist = [w for w in wsj if re.search('^[0-9]+\.[0-9]+$', w)]
print(wordlist[:10])

* -는 사이에 있는 것 모두, 즉 0-9는 0,1,2,3,4,5,6,7,8,9를 말한다. 
* 이 경우 조건은 모든 소수를 의미할 것.

In [ ]:
wordlist = [w for w in wsj if re.search('^[A-Z]+\$$', w)]
print(wordlist[:10])

* $는 ~로 끝나는 것을 의미한다고 말했는데, 정말 글자 그대로의 $가 쓰고 싶다면 앞에\를 붙이면 된다.
* 이 경우에는 대문자 알파벳으로 시작하고, $로 끝나는 것을 보여주라는 의미 

In [ ]:
wordlist = [w for w in wsj if re.search('^[0-9]{4}$', w)]
print(wordlist[:10])

In [ ]:
wordlist = [w for w in wsj if re.search('^[a-z]{5,}-[a-z]{2,3}-[a-z]{,6}$', w)]
print(wordlist[:10])

In [ ]:
wordlist = [w for w in wsj if re.search('(ed|ing)$', w)]
print(wordlist[:10])

In [3]:
word = 'supercalifragilisticexpialidocious'
result = re.findall(r'[aeiou]', word)
print(result)

['u', 'e', 'a', 'i', 'a', 'i', 'i', 'i', 'e', 'i', 'a', 'i', 'o', 'i', 'o', 'u']


* 주어진 word에서 모음을 찾아라

In [ ]:
result = re.findall('[aeiou](..)[aeiou]', word)
print(result)

* 주어진 word에 모음과 모음 사이에 자음 두 글자가 들어가 있다면, 이 자음 두 개의 조합을 모두 보여주라는 의미

In [5]:
result = re.findall('[^aeiou].+[^aeiou]', word)
print(result)

['supercalifragilisticexpialidocious']


* ^는 not을 의미. 이 경우에는 모음이 아닌 것을 의미
* .+ 캐릭터 하나 이상 - 모음에 다른 거 하나에 모음이 있어야 한다.
* greedy search : 내부적으로 포함되는 것은 무시, 가장 큰 단위로 찾는다

In [6]:
result = re.findall('[^aeiou].+?[^aeiou]', word)
print(result)

['sup', 'rcal', 'frag', 'lis', 'tic', 'xpial', 'doc']


* reluctant search : ?가 포함되어 있는 것으로, 가장 좁은 단위로 찾아라

In [7]:
wsj = sorted(set(nltk.corpus.treebank.words()))


* word street journal
* set은 집합으로 만드는 것, 중복 등을 제거하고 사전처럼 만드는 작업.

In [9]:
fdist = nltk.FreqDist(vs for word in wsj for vs in re.findall(r'[aeiou]{2,}', word))
fdist.most_common(10)

[('io', 549),
 ('ea', 476),
 ('ie', 331),
 ('ou', 329),
 ('ai', 261),
 ('ia', 253),
 ('ee', 217),
 ('oo', 174),
 ('ua', 109),
 ('au', 106)]

* findall은 뒤의 string에서 있는 것을 다 찾아라
* Freqdist의 variable type은 list. 
* [aeiou]{2,} = aeiou가 2개 이상 들어 있기만 하면 그것을 찾아와서 list로 만들어라. 
* 그 list에서 fdist를 받고 . most_common 하면, 상위 10개를 보여준다.
* 'io', 549 는 io가 549번 나와서 가장 자주 나왔다는 것.

In [13]:
wordlist = [vs for word in wsj for vs in re.findall('[aeiou]{2,}',word)]

* word에서 조건에 해당하는 단어들로 새로운 list를, wordlist를 만들어라.
* 모음이 2번 이상 연속되는 것을 찾아서 vs의 리스트, wordlist로 만들어라.

In [14]:
wordlist

['ea',
 'oi',
 'ea',
 'ou',
 'oi',
 'ea',
 'ea',
 'oi',
 'oi',
 'ea',
 'io',
 'ea',
 'ea',
 'ea',
 'oi',
 'ea',
 'ea',
 'ea',
 'ea',
 'ea',
 'ea',
 'ea',
 'ee',
 'ea',
 'ea',
 'ea',
 'ea',
 'ea',
 'ea',
 'ea',
 'ea',
 'oi',
 'ea',
 'ea',
 'ou',
 'ou',
 'ou',
 'ie',
 'ui',
 'io',
 'ua',
 'io',
 'ai',
 'ai',
 'ai',
 'io',
 'ie',
 'ue',
 'ue',
 'ia',
 'ie',
 'ea',
 'ai',
 'ou',
 'ia',
 'ei',
 'ie',
 'ea',
 'ea',
 'ie',
 'ia',
 'ia',
 'ua',
 'ie',
 'io',
 'ea',
 'ia',
 'io',
 'ui',
 'ia',
 'ia',
 'ea',
 'iai',
 'ai',
 'ia',
 'ia',
 'ia',
 'ia',
 'ia',
 'io',
 'oo',
 'io',
 'ia',
 'ia',
 'ia',
 'ia',
 'ue',
 'ea',
 'ai',
 'ai',
 'ue',
 'ie',
 'au',
 'ea',
 'ea',
 'ea',
 'ea',
 'eau',
 'au',
 'ei',
 'ei',
 'ei',
 'ei',
 'ei',
 'ia',
 'ie',
 'io',
 'ue',
 'oa',
 'oei',
 'oe',
 'ia',
 'oo',
 'oo',
 'oo',
 'eau',
 'ou',
 'ou',
 'ai',
 'ou',
 'ai',
 'oo',
 'ea',
 'au',
 'ia',
 'ea',
 'ea',
 'ee',
 'ia',
 'ai',
 'oa',
 'oo',
 'oo',
 'oo',
 'ei',
 'ei',
 'ea',
 'ui',
 'ui',
 'eau',
 'ie',
 'ia',
 

In [15]:
fdist = nltk.FreqDist(wordlist)
fdist.most_common(10)

[('io', 549),
 ('ea', 476),
 ('ie', 331),
 ('ou', 329),
 ('ai', 261),
 ('ia', 253),
 ('ee', 217),
 ('oo', 174),
 ('ua', 109),
 ('au', 106)]

* (wordlist)자리에 사전을 넣으면, 빈도수가 모두 1로 나온다. 사전은 중복이 없기 때문이다. 

In [16]:
english_udhr = nltk.corpus.udhr.words('English-Latin1')
print(english_udhr)

['Universal', 'Declaration', 'of', 'Human', 'Rights', ...]


In [18]:
regexp = r'^[AEIOUaeiou]+|[AEIOUaeiou]+$|[^AEIOUaeiou]'
def compress(word):
    pieces = re.findall(regexp, word)
    return ''.join(pieces)
[compress(w) for w in english_udhr [:75]]


['Unvrsl',
 'Dclrtn',
 'of',
 'Hmn',
 'Rghts',
 'Prmble',
 'Whrs',
 'rcgntn',
 'of',
 'the',
 'inhrnt',
 'dgnty',
 'and',
 'of',
 'the',
 'eql',
 'and',
 'inlnble',
 'rghts',
 'of',
 'all',
 'mmbrs',
 'of',
 'the',
 'hmn',
 'fmly',
 'is',
 'the',
 'fndtn',
 'of',
 'frdm',
 ',',
 'jstce',
 'and',
 'pce',
 'in',
 'the',
 'wrld',
 ',',
 'Whrs',
 'dsrgrd',
 'and',
 'cntmpt',
 'fr',
 'hmn',
 'rghts',
 'hve',
 'rsltd',
 'in',
 'brbrs',
 'acts',
 'whch',
 'hve',
 'outrgd',
 'the',
 'cnscnce',
 'of',
 'mnknd',
 ',',
 'and',
 'the',
 'advnt',
 'of',
 'a',
 'wrld',
 'in',
 'whch',
 'hmn',
 'bngs',
 'shll',
 'enjy',
 'frdm',
 'of',
 'spch',
 'and']

* 함수 : 입력을 받아서 내부적으로 바꾸고 출력해준다.
* def = define 
* def compress(word) : '이 함수는 compress다'라는 arbitrary한 정의. ()안이 입력이다.
* 입력 = word, 출력 = return 뒷부분

* 이 경우 def부터 return까지 세 줄이 함수다.
* | bar은 or을 의미한다. 조건들 중에 하나라도 만족시키면 찾아라.
* ^[] (시작되는 것 지정)과 [^ ] (not~)을 헷갈리지 않도록 주의할 것.
* 이 경우 모음으로 시작하거나, 모음으로 끝나거나, 자음이 들어가면 찾아라.
* 출력이 join, join에 list가 들어오는데 이 list들을 ''으로, 즉 공백없이 붙여라. 


* 결과값으로는 시작, 끝에 들어간 모음을 제외한 나머지 모음을 삭제한 단어들이 나옴.

In [19]:
rotokas_words = nltk.corpus.toolbox.words('rotokas.dic')


In [22]:
cvs = [cv for w in rotokas_words for cv in re.findall(r'[ptksvr][aeiou]', w)]
print(cvs)

['ka', 'ka', 'ka', 'ka', 'ka', 'ro', 'ka', 'ka', 'vi', 'ko', 'ka', 'ka', 'vo', 'ka', 'ka', 'ko', 'ka', 'ka', 'si', 'ka', 'ka', 'ka', 'ka', 'ko', 'ka', 'ki', 'to', 'ka', 'ku', 'pa', 'to', 'ka', 'va', 'ka', 'pa', 'ka', 'pe', 'ka', 'pi', 'ka', 'pi', 'ka', 'pi', 'pa', 'to', 'ka', 'pi', 'si', 'ka', 'pi', 'si', 'vi', 'ra', 'ka', 'po', 'ka', 'po', 'pa', 'to', 'ka', 'ra', 'ka', 're', 'ka', 're', 'ko', 'ka', 're', 'ko', 'pi', 'ka', 're', 'to', 're', 'va', 'ka', 'va', 'ka', 'va', 'ka', 've', 'ka', 'ka', 've', 'ka', 'pi', 'ka', 've', 'ka', 'pi', 'vi', 'ra', 'ka', 've', 'ka', 'vi', 'ra', 'ka', 'ka', 'ka', 'ka', 'ka', 'ka', 'ka', 'ka', 'ro', 'ka', 'ka', 'ka', 'ka', 'so', 'to', 'ka', 'ka', 'vi', 'ra', 'ka', 'ke', 'ru', 'ka', 'pa', 'ka', 'pi', 'ka', 'pi', 'ka', 'pi', 'vi', 'ra', 'ka', 're', 'si', 'ka', 're', 'si', 'vi', 'ra', 'ka', 'tu', 'ka', 'tu', 'pi', 'ka', 'tu', 'vi', 'ra', 'ka', 'vi', 'ro', 'ka', 've', 'ka', 'ka', 'ka', 'ka', 'ka', 'po', 'ri', 'ka', 'po', 'ri', 'pi', 'ka', 'po', 'ri', 'vi', 'ra

In [23]:
cfd = nltk.ConditionalFreqDist(cvs) # CFD works pairwise only
cfd.tabulate()

    a   e   i   o   u 
k 418 148  94 420 173 
p  83  31 105  34  51 
r 187  63  84  89  79 
s   0   0 100   2   1 
t  47   8   0 148  37 
v  93  27 105  48  49 


* number of occurences을 conditional하게 보여준다. 

In [ ]:
cv_word_pairs = [(cv, w) for w in rotokas_words for cv in re.findall(r'[ptksvr][aeiou]', w)]
cv_index = nltk.Index(cv_word_pairs)
cv_index['po']